### PR TITLE
[Interaction.Scroll] - Initial Implementation

### PR DESCRIPTION
**Pull request by bluong**
Wraps the dispatchers return values to ensure that the deltaY is returned in pixel space.

NOTE: I cannot really think of much capability of this since we are currently using it for the usecase of zooming.  At the very least, I do want consumers to be able to interact with their components with their mouse wheel.

NOTE: This is also prefacing the idea of zooming which will go down a similar path as pan.

### CODE REVIEW

File: src/interactions/scrollInteraction.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+      super._anchor(component, hitBox);
+      this._mouseDispatcher = Dispatcher.Mouse.getDispatcher(<SVGElement> this._componentToListenTo.content().node());
+      this._mouseDispatcher.onWheel("Interaction.Scroll" + this.getID(), (p: Point, e: WheelEvent) => this._handleScrollEvent(p, e));
+    }
+
+    private _handleScrollEvent(p: Point, e: WheelEvent) {
+      var translatedP = this._translateToComponentSpace(p);
+      if (this._isInsideComponent(translatedP)) {
+        if (this._scrollCallback) {
+          var deltaPixelAmount = e.deltaY * (e.deltaMode ? 120 : 1);
```
Comment (jtlan) [inline src/interactions/scrollInteraction.ts]: Is there testing for this functionality?

Also it'd be good if we could find a reference for this 120...

**Comment by jtlan** [comment]
Open discussion:

Should we type-define callback functions?
**arguments for**
* Distinguishes callbacks from functions with the same signature
* Allows documentation of callback parameters and return types

**arguments against**
* Methods passed or returning the callback type are hard to use without cross-referencing the callback type definition

File: src/interactions/scrollInteraction.ts (hunk 1 above)
Comment (bluong) [inline src/interactions/scrollInteraction.ts]: I looked through the d3 js source and found the 120 there with pretty much this exact line of code
`e.deltaY * (e.deltaMode ? 120 : 1);`

File: src/interactions/scrollInteraction.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+export module Interaction {
+
+  export type ScrollCallback = (p: Point, deltaAmount: number) => any;
+
+  export class Scroll extends Interaction.AbstractInteraction {
+
+    /**
+     * The number of pixels occupied in a line.
+     */
+    public static PIXELS_PER_LINE = 120;
```
Comment (jtlan) [inline src/interactions/scrollInteraction.ts]: Can we find a source for this value?

**Comment by jtlan** [comment]
Since it's called `Interaction.Scroll`, should we also capture non-wheel scroll events, such as a two-finger scroll gesture?

File: plottable.js [hunk 3]
```diff
@@ -10261,6 +10261,55 @@ var Plottable;
...
+                _super.prototype._anchor.call(this, component, hitBox);
+                this._mouseDispatcher = Plottable.Dispatcher.Mouse.getDispatcher(this._componentToListenTo.content().node());
+                this._mouseDispatcher.onWheel("Interaction.Scroll" + this.getID(), function (p, e) { return _this._handleScrollEvent(p, e); });
+            };
+            Scroll.prototype._handleScrollEvent = function (p, e) {
+                var translatedP = this._translateToComponentSpace(p);
+                if (this._isInsideComponent(translatedP)) {
+                    if (this._scrollCallback) {
+                        e.preventDefault();
+                        var deltaPixelAmount = e.deltaY * (e.deltaMode ? Scroll.PIXELS_PER_LINE : 1);
```
Comment (jtlan) [inline plottable.js]: Aren't there three delta modes? https://developer.mozilla.org/en-US/docs/Web/Events/wheel

File: src/interactions/scrollInteraction.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+      this._mouseDispatcher = Dispatcher.Mouse.getDispatcher(<SVGElement> this._componentToListenTo.content().node());
+      this._mouseDispatcher.onWheel("Interaction.Scroll" + this.getID(), (p: Point, e: WheelEvent) => this._handleScrollEvent(p, e));
+    }
+
+    private _handleScrollEvent(p: Point, e: WheelEvent) {
+      var translatedP = this._translateToComponentSpace(p);
+      if (this._isInsideComponent(translatedP)) {
+        if (this._scrollCallback) {
+          e.preventDefault();
+          var deltaPixelAmount = e.deltaY * (e.deltaMode ? Scroll.PIXELS_PER_LINE : 1);
```
Comment (jtlan) [inline src/interactions/scrollInteraction.ts]: Aren't there three delta modes? https://developer.mozilla.org/en-US/docs/Web/Events/wheel

Comment (bluong) [inline src/interactions/scrollInteraction.ts]: While true, I've looked around implementations and only saw the 2 modes being cased out... I would need the PIXELS_PER_PAGE to switch off

File: src/interactions/scrollInteraction.ts (hunk 2 above)
Comment (bluong) [inline src/interactions/scrollInteraction.ts]: The best I've got is this code from D3
`-d3.event.deltaY * (d3.event.deltaMode ? 120 : 1);`

I'll go look around for a more legitimate source

Comment (bluong) [inline src/interactions/scrollInteraction.ts]: Based on this post http://stackoverflow.com/questions/20110224/what-is-the-height-of-a-line-in-a-wheel-event-deltamode-dom-delta-line

it's not consistent.  Thus, D3 is basically using a guess estimate.  We can stick with the estimate to get behavior similar to D3 or we can come up with our own logic.

**Comment by bluong** [comment]
This does not seem relevant anymore since it will be rolled to be an internal thing inside PanZoom

**Pull request closed without merging**